src/generator.cs

```diff
@@ -1117,6 +1118,145 @@ static string FindProtocolInterface (Type type, MemberInfo pi)
+			temp = string.Format ("{0} == null ? null : ", pi != null ? pi.Name.GetSafeParamName () : "value");
+
+		if (className == "NSNumber")
+			temp = string.Format ("new NSNumber ({1}{0});", isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
+
+		else if (className == "NSValue") {
+			switch (retType.Name) {
+			case "CATransform3D":
+			case "CGAffineTransform":
+			case "CGPoint":
+			case "CGRect":
+			case "CGSize":
+			case "CGVector":
+			case "CMTimeMapping":
+			case "CMTimeRange":
+			case "CMTime":
+			case "MKCoordinateSpan":
+			case "CLLocationCoordinate2D":
+			case "PointF":
+			case "RectangleF":
+			case "SCNMatrix4":
+			case "SizeF":
+			case "UIEdgeInsets":
+			case "UIOffset":
+				temp = string.Format ("NSValue.From{0} ({{1}}{{0}});", retType.Name);
+				break;
+			case "SCNVector3":
+			case "SCNVector4":
+				temp = "NSValue.FromVector ({1}{0});";
+				break;
+			case "NSRange":
+				temp = "NSValue.FromRange ({1}{0});";
+				break;
+			default:
+				throw new BindingException (1049, true, "Could not box type {0} into {1} container used on {2} member decorated with [BindAs].", retType.Name, "NSValue", minfo?.mi?.Name ?? pi?.Name);
+			}
+
+			temp = string.Format (temp, isNullable ? ".Value" : string.Empty, pi != null ? pi.Name.GetSafeParamName () : "value");
+		} else if (isValueType) {
+			// Magic for enums/smart enums
+		} else
+			throw new BindingException (1048, true, "Unsupported type {0} decorated with [BindAs]", retType.Name);
+
+		return temp;
+	}
+
+	static Dictionary<Type,string> NSNumberReturnMap = new Dictionary<Type, string> {
+		{ typeof (bool), ".BoolValue" }, { typeof (byte), ".ByteValue" }, { typeof (double), ".DoubleValue" },
+		{ typeof (float), ".FloatValue" }, { typeof (short), ".Int16Value" }, { typeof (int), ".Int32Value" },
+		{ typeof (long), ".Int64Value" }, { typeof (sbyte), ".SByteValue" }, { typeof (ushort), ".UInt16Value" },
+		{ typeof (uint), ".UInt32Value" }, { typeof (ulong), ".UInt64Value" },
+#if XAMCORE_2_0
+		{ typeof (nfloat), ".NFloatValue" }, { typeof (nint), ".NIntValue" }, { typeof (nuint), ".NUIntValue" },
+#endif
+	};
+
+	string GetFromBindAsWrapper (MemberInformation minfo)
+	{
+		var attrib = GetBindAsAttribute (minfo.mi);
+		var retType = Nullable.GetUnderlyingType (attrib.Type) ?? attrib.Type;
+		var isValueType = retType.IsValueType;
+		var append = string.Empty;
+		var property = minfo.mi as PropertyInfo;
+		var method = minfo.mi as MethodInfo;
+
+		if (method?.ReturnType?.Name == "NSNumber" || property?.PropertyType?.Name == "NSNumber") {
+			if (!NSNumberReturnMap.TryGetValue (retType, out append))
+				throw new BindingException (1049, true, "Could not unbox type {0} from {1} container used on {2} member decorated with [BindAs].", retType.Name, "NSNumber", minfo.mi.Name);
+
+		} else if (method?.ReturnType.Name == "NSValue" || property?.PropertyType.Name == "NSValue") {
+			switch (retType.Name) {
+			case "CATransform3D":
```

Comment:
I don't really like doing string comparison for type checks, since it breaks if the customer tries to declare (and use) his own types with the same names (which is obviously a bad idea, but people do the strangest things).

I did like the `NSNumberReturnMap` dictionary though, maybe you can do something similar for `NSValue` as well? The same thing is also happening in `GetToBindAsWrapper`.